gap/lqueue.gi

```diff
@@ -1,174 +1,203 @@
-      tail := queue![QTAIL];
-      queue![QDATA][tail] := el;
-      tail := tail + 1;
-      queue![QTAIL] := tail;
+function(queue,el)
+    local head, tail, last;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # Special case for empty queue,
+    # head = tail and queue![QDATA][tail] is
+    # not bound
+    if not IsBound(queue![QDATA][tail]) then
+        queue![QDATA][tail] := el;
     else
-      queue![QTAIL] := 1;
-      queue![QDATA][last] := el;
+        if tail = last then
+            tail := 1;
+        else
+            tail := tail + 1;
+        fi;
+        queue![QDATA][tail] := el;
+        queue![QTAIL] := tail;
+
+        # If queue is full expand
+        if ((head = 1) and (tail = last))
+           or (tail + 1 = head) then
+            PlistQueueExpand(queue);
+        fi;
     fi;
-  elif tail + 1 <> head then
-    queue![QDATA][tail] := el;
-    tail := tail + 1;
-    queue![QTAIL] := tail;
-  else
-    PlistQueueExpand(queue);
-    tail := queue![QTAIL];
-    queue![QDATA][tail] := el;
-    tail := tail + 1;
-    queue![QTAIL] := tail;
-  fi;
 end);
 
 InstallGlobalFunction(PlistQueuePushFront,
 function(queue, el)
-  local head, tail, last;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head = 1 then
-    if tail = last then
-      PlistQueueExpand(queue);
-      head := queue![QCAPACITY];
-      queue![QDATA][head] := el;
-      queue![QHEAD] := head;
+    local head, tail, last;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # If the queue is empty
+    if not IsBound(queue![QDATA][head]) then
+        queue![QDATA][head] := el;
     else
-      queue![QHEAD] := last;
-      queue![QDATA][last] := el;
+        if head = 1 then
+            head := last;
+        else
+            head := head - 1;
+        fi;
+        queue![QDATA][head] := el;
+        queue![QHEAD] := head;
+
+        if (head = 1 and tail = last)
+           or (tail + 1 = head) then
+            PlistQueueExpand(queue);
+        fi;
     fi;
-  elif tail + 1 <> head then
-    head := head - 1;
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  else
-    PlistQueueExpand(queue);
-    head := Size(queue);
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  fi;
 end);
 
 InstallGlobalFunction(PlistQueuePopFront,
 function(queue)
-  local head, tail, last, result;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head <> tail then
-    if head = last then
-      head := 1;
-      result := queue![QDATA][last];
-      queue![QDATA][last] := fail;
-    else
-      head := head + 1;
-      result := queue![QDATA][head-1];
-      queue![QDATA][head-1] := fail;
+    local head, tail, last, result;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # TODO: What if someone wants to store fails?
```

Comment:
That's something they currently can't in e.g. hash maps, too. That's a general decision we could/should make: How to react to a pop/lookup when there is nothing? Right now, we return `fail`, and prohibit storing `fail` in data structures (well... hash maps do -- the `Push` operations here don't check for `fail`, and thus allow it in, which I'd consider a bug).

The alternative would be to use `Error` in that case. I'd also be happy with that. I.e. if somebody does a pop or peek that fails, call `Error`. Then it is up to the caller to first check `Size` / `IsEmpty` / `IsBound(...)` / ...